src/python/grpcio/grpc/_interceptor.py

```diff
@@ -235,8 +236,8 @@ def continuation(new_details, request):
                     credentials=new_credentials,
                     wait_for_ready=new_wait_for_ready)
                 return _UnaryOutcome(response, call)
-            except grpc.RpcError as rpc_error:
-                return rpc_error
+            except _Rendezvous as rendezvous:
```

Comment:
Can we use a more generic interface here, like `grpc.Call`? It should return a client-side context object.

Reply:
We cannot, really, because we need an object to properly implement both `Call` and `RpcError` interfaces properly, and it would effectively be a `_Renzdevous` in practice. Alternatively, we could simply omit that branch, as it is just an optimization for the common case of returning a `_Renzdevous`.

Reply:
Don't we also need to implement the `Future` interface to actually have the requisite `exception` attribute? The doc says:

> In the event of RPC completion, the return Call-Future’s result value will be the response message of the RPC. Should the event terminate with non-OK status, the returned Call-Future’s exception value will be an RpcError.

Omitting the special casing altogether looks like it would lead to swallowing the rpc error code, as mentioned by the reviewer on your original change [here](https://github.com/grpc/grpc/pull/14639#pullrequestreview-102314676).

The docs on `with_call` allow it to raise an `RpcError` which will also be a `Call`, which we must then ensure is also a `Future` when we return it here. I am fully willing to believe the current implementation will/should always result in a `_Rendezvous` here, but we should program to the API, not the (underscore-prefixed) `_Rendezvous` implementation.

This code and the potential for type errors here definitely warrant a test. We can take over (PR-as-issue?) but it would be helpful when prioritizing this for us to know if you're sending this change to fix a known breakage (PR-as-bug) or if this is purely intended to future proof the code in correcting an oversight from the last change (PR-as-P2).

Reply:
@ericgribkoff There's definitely a bug there that might bite other users, but if what you are asking is whether I encountered it in real usage, the answer is no; just realized that the code is definitely buggy. There might be other bugs or a better way to fundamentally resolve it. There's definitely complexity and room for error that's rooted in `_Rendezvous`'s plethora of interfaces that is hard to understand or enforce without proper typechecking.  I agree that proper testing is warranted, perhaps with clarification on the documentation if there's anything missing.  Looks like `grpc.Future` is indeed necessary as well, so an adapter class might be needed here. That said, special casing `_Rendezvous` as an *optimization* so that you don't construct intermediate classes in the common case might still be a valid thing to do. Please feel free to do what's best in your opinion to resolve (the whole thing is still marked as experimental, so even minor design changes may still be appropriate).